csharp/src/Google.Protobuf/ParsingPrimitives.cs

```diff
@@ -638,20 +604,76 @@ public static string ReadRawString(ref ReadOnlySpan<byte> buffer, ref ParserInte
             }
 #endif
 
-            var decoder = WritingPrimitives.Utf8Encoding.GetDecoder();
+            return ReadStringSlow(ref buffer, ref state, length);
+        }
+
+        /// <summary>
+        /// Reads a string assuming that it is spread across multiple spans in a <see cref="ReadOnlySequence{T}"/>.
+        /// </summary>
+        private static string ReadStringSlow(ref ReadOnlySpan<byte> buffer, ref ParserInternalState state, int length)
+        {
+            ValidateCurrentLimit(ref buffer, ref state, length);
```

Comment:
I think ValidateCurrentLimit invocation should be in ReadRawString (in the intial checking of preconditions) - at least that would make it consistent with ReadRawBytes and SkipRawBytes - and I also think like this we're still missing one possible scenario (entire string fits into the remaining current buffer but the length is still over the currentLimit) and I think that's a bug (not introduced by this PR, it existed even before this PR it seems, but still).

Let's move `ValidateCurrentLimit(ref buffer, ref state, length);` to line ~586 (just after the negative size check)?

Reply:
> and I also think like this we're still missing one possible scenario (entire string fits into the remaining current buffer but the length is still over the currentLimit) and I think that's a bug

bufferPos is limited by currentLimit. If content fits inside bufferPos then it will also be inside the limit.